Add terraform static security analysis to PR checks

~~We should add terraform-compliance checks to our pipeline: https://github.com/eerkunt/terraform-compliance, wdyt @hferentschik?
One use case would be to check if created resources have tags or not: https://terraform-compliance.com/pages/Examples/tags_related.html (This will help us to fix some issues with terraform destroy)~~

Seems like tfsec also does similar things: https://github.com/liamg/tfsec.
Running tfsec on the module shows some issues:
```
9 potential problems detected:

Problem 1

  [AWS002][ERROR] Resource 'aws_s3_bucket.logs_jenkins_x' does not have logging enabled.
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:5-13

       2 | // Create the AWS S3 buckets for Long Term Storage based on flags
       3 | // See https://www.terraform.io/docs/providers/aws/r/s3_bucket.html
       4 | // ----------------------------------------------------------------------------
       5 | resource "aws_s3_bucket" "logs_jenkins_x" {
       6 |   count         = var.enable_logs_storage ? 1 : 0
       7 |   bucket_prefix = "logs-${var.cluster_name}-"
       8 |   acl           = "private"
       9 |   tags = {
      10 |     Owner = "Jenkins-x"
      11 |   }
      12 |   force_destroy = var.force_destroy
      13 | }
      14 | 
      15 | resource "aws_s3_bucket" "reports_jenkins_x" {
      16 |   count         = var.enable_reports_storage ? 1 : 0

  See https://github.com/liamg/tfsec/wiki/AWS002 for more information.

Problem 2

  [AWS017][ERROR] Resource 'aws_s3_bucket.logs_jenkins_x' defines an unencrypted S3 bucket (missing server_side_encryption_configuration block).
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:5-13

       2 | // Create the AWS S3 buckets for Long Term Storage based on flags
       3 | // See https://www.terraform.io/docs/providers/aws/r/s3_bucket.html
       4 | // ----------------------------------------------------------------------------
       5 | resource "aws_s3_bucket" "logs_jenkins_x" {
       6 |   count         = var.enable_logs_storage ? 1 : 0
       7 |   bucket_prefix = "logs-${var.cluster_name}-"
       8 |   acl           = "private"
       9 |   tags = {
      10 |     Owner = "Jenkins-x"
      11 |   }
      12 |   force_destroy = var.force_destroy
      13 | }
      14 | 
      15 | resource "aws_s3_bucket" "reports_jenkins_x" {
      16 |   count         = var.enable_reports_storage ? 1 : 0

  See https://github.com/liamg/tfsec/wiki/AWS017 for more information.

Problem 3

  [AWS002][ERROR] Resource 'aws_s3_bucket.reports_jenkins_x' does not have logging enabled.
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:15-23

      12 |   force_destroy = var.force_destroy
      13 | }
      14 | 
      15 | resource "aws_s3_bucket" "reports_jenkins_x" {
      16 |   count         = var.enable_reports_storage ? 1 : 0
      17 |   bucket_prefix = "reports-${var.cluster_name}-"
      18 |   acl           = "private"
      19 |   tags = {
      20 |     Owner = "Jenkins-x"
      21 |   }
      22 |   force_destroy = var.force_destroy
      23 | }
      24 | 
      25 | resource "aws_s3_bucket" "repository_jenkins_x" {
      26 |   count         = var.enable_repository_storage ? 1 : 0

  See https://github.com/liamg/tfsec/wiki/AWS002 for more information.

Problem 4

  [AWS017][ERROR] Resource 'aws_s3_bucket.reports_jenkins_x' defines an unencrypted S3 bucket (missing server_side_encryption_configuration block).
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:15-23

      12 |   force_destroy = var.force_destroy
      13 | }
      14 | 
      15 | resource "aws_s3_bucket" "reports_jenkins_x" {
      16 |   count         = var.enable_reports_storage ? 1 : 0
      17 |   bucket_prefix = "reports-${var.cluster_name}-"
      18 |   acl           = "private"
      19 |   tags = {
      20 |     Owner = "Jenkins-x"
      21 |   }
      22 |   force_destroy = var.force_destroy
      23 | }
      24 | 
      25 | resource "aws_s3_bucket" "repository_jenkins_x" {
      26 |   count         = var.enable_repository_storage ? 1 : 0

  See https://github.com/liamg/tfsec/wiki/AWS017 for more information.

Problem 5

  [AWS002][ERROR] Resource 'aws_s3_bucket.repository_jenkins_x' does not have logging enabled.
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:25-33

      22 |   force_destroy = var.force_destroy
      23 | }
      24 | 
      25 | resource "aws_s3_bucket" "repository_jenkins_x" {
      26 |   count         = var.enable_repository_storage ? 1 : 0
      27 |   bucket_prefix = "repository-${var.cluster_name}-"
      28 |   acl           = "private"
      29 |   tags = {
      30 |     Owner = "Jenkins-x"
      31 |   }
      32 |   force_destroy = var.force_destroy
      33 | }
      34 | 

  See https://github.com/liamg/tfsec/wiki/AWS002 for more information.

Problem 6

  [AWS017][ERROR] Resource 'aws_s3_bucket.repository_jenkins_x' defines an unencrypted S3 bucket (missing server_side_encryption_configuration block).
  /home/ankitm123/work/terraform-aws-eks-jx/modules/cluster/storage.tf:25-33

      22 |   force_destroy = var.force_destroy
      23 | }
      24 | 
      25 | resource "aws_s3_bucket" "repository_jenkins_x" {
      26 |   count         = var.enable_repository_storage ? 1 : 0
      27 |   bucket_prefix = "repository-${var.cluster_name}-"
      28 |   acl           = "private"
      29 |   tags = {
      30 |     Owner = "Jenkins-x"
      31 |   }
      32 |   force_destroy = var.force_destroy
      33 | }
      34 | 

  See https://github.com/liamg/tfsec/wiki/AWS017 for more information.

Problem 7

  [AWS002][ERROR] Resource 'aws_s3_bucket.vault-unseal-bucket' does not have logging enabled.
  /home/ankitm123/work/terraform-aws-eks-jx/modules/vault/main.tf:28-38

      25 | // Vault S3 bucket
      26 | // See https://www.terraform.io/docs/providers/aws/r/s3_bucket.html
      27 | // ----------------------------------------------------------------------------
      28 | resource "aws_s3_bucket" "vault-unseal-bucket" {
      29 |   bucket_prefix = "vault-unseal-${var.cluster_name}-"
      30 |   acl           = "private"
      31 |   tags = {
      32 |     Name        = "Vault unseal bucket"
      33 |   }
      34 |   versioning {
      35 |     enabled = false
      36 |   }
      37 |   force_destroy = var.force_destroy
      38 | }
      39 | 
      40 | // ----------------------------------------------------------------------------
      41 | // Vault DynamoDB Table

  See https://github.com/liamg/tfsec/wiki/AWS002 for more information.

Problem 8

  [AWS017][ERROR] Resource 'aws_s3_bucket.vault-unseal-bucket' defines an unencrypted S3 bucket (missing server_side_encryption_configuration block).
  /home/ankitm123/work/terraform-aws-eks-jx/modules/vault/main.tf:28-38

      25 | // Vault S3 bucket
      26 | // See https://www.terraform.io/docs/providers/aws/r/s3_bucket.html
      27 | // ----------------------------------------------------------------------------
      28 | resource "aws_s3_bucket" "vault-unseal-bucket" {
      29 |   bucket_prefix = "vault-unseal-${var.cluster_name}-"
      30 |   acl           = "private"
      31 |   tags = {
      32 |     Name        = "Vault unseal bucket"
      33 |   }
      34 |   versioning {
      35 |     enabled = false
      36 |   }
      37 |   force_destroy = var.force_destroy
      38 | }
      39 | 
      40 | // ----------------------------------------------------------------------------
      41 | // Vault DynamoDB Table

  See https://github.com/liamg/tfsec/wiki/AWS017 for more information.

Problem 9

  [AWS019][WARNING] Resource 'aws_kms_key.kms_vault_unseal' does not have KMS Key auto-rotation enabled.
  /home/ankitm123/work/terraform-aws-eks-jx/modules/vault/main.tf:71-92

      68 | // Vault KMS Key
      69 | // See https://www.terraform.io/docs/providers/aws/r/kms_key.html
      70 | // ----------------------------------------------------------------------------
      71 | resource "aws_kms_key" "kms_vault_unseal" {
      72 |   description   = "KMS Key for bank vault unseal"
      73 |   policy        = <<POLICY
      74 | {
      75 |     "Version": "2012-10-17",
      76 |     "Statement": [
      77 |         {
      78 |             "Sid": "EnableIAMUserPermissions",
      79 |             "Effect": "Allow",
      80 |             "Principal": {
      81 |                 "AWS": [
      82 |                     "${data.aws_iam_user.vault_user.arn}",
      83 |                     "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
      84 |                 ]
      85 |             },
      86 |             "Action": "kms:*",
      87 |             "Resource": "*"
      88 |         }
      89 |     ]
      90 | }
      91 | POLICY
      92 | }
      93 | 
      94 | // ----------------------------------------------------------------------------
      95 | // Permissions that will need to be attached to the provides IAM Username

  See https://github.com/liamg/tfsec/wiki/AWS019 for more information.

```
We should fix them asap.

--- a/.github/issue_template.md
+++ b/.github/issue_template.md
@@ -1,0 +1,27 @@
+### Summary 
+
+### Steps to reproduce the behavior
+
+### Expected behavior
+
+### Actual behavior
+
+### Terraform version
+
+The output of `terraform version` is:
+
+```
+COPY OUTPUT HERE
+```
+
+### Module version
+
+<!--
+What version of the module are u using?
+-->
+
+### Operating system
+
+<!--
+In which OS are you running the terraform apply?
+-->

--- a/.github/pull_request_template.md
+++ b/.github/pull_request_template.md
@@ -1,0 +1,17 @@
+<!--
+Add this section after we add tests and compliance
+#### Submitter checklist
+
+- [ ] Change is code complete and matches issue description.
+- [ ] Change is covered by existing or new tests.
+- [ ] Readme and jx-docs (https://jenkins-x.io/docs/install-setup/installing/create-cluster/eks/) are updated 
+-->
+#### Description
+
+
+#### Special notes for the reviewer(s)
+
+
+#### Which issue this PR fixes
+
+fixes #

--- a/.gitignore
+++ b/.gitignore
@@ -36,7 +36,9 @@ override.tf.json
 jx-requirements.yml
 kubeconfig*
 
+# Misc
 .DS_Store
+tfsec
 
 #Intellij
 .idea

--- a/Makefile
+++ b/Makefile
@@ -1,3 +1,15 @@
+ifeq ($(OS),Windows_NT)
+    OS := windows
+else
+    UNAME_S := $(shell uname -s)
+    ifeq ($(UNAME_S),Linux)
+        OS := linux
+    endif
+    ifeq ($(UNAME_S),Darwin)
+        OS := darwin
+    endif
+endif
+
 SHELL_SPEC_DIR ?= /var/tmp
 
 .DEFAULT_GOAL := help
@@ -28,15 +40,15 @@ lint: init ## Verifies Terraform syntax
 fmt: ## Reformats Terraform files accoring to standard
 	terraform fmt
 
-install-tfsec: ##Installs tfsec
-	curl -L "$$(curl -s https://api.github.com/repos/liamg/tfsec/releases/latest | grep -o -E "https://.+?-linux-amd64")" > tfsec;\
+tfsec: ## Installs tfsec
+	curl -L "$$(curl -s https://api.github.com/repos/liamg/tfsec/releases/latest | grep -o -E "https://.+?-$(OS)-amd64(.exe)?")" > tfsec;\
 	chmod +x ./tfsec
 
-check-tfsec: # check if tfsec is installed
+check-tfsec: ## Check if tfsec is installed
 	./tfsec --version
 
-.PHONY: tfsec
-tfsec: install-tfsec check-tfsec #Runs tfsec
+.PHONY: run-tfsec
+run-tfsec: tfsec check-tfsec ## Runs tfsec
 	./tfsec . -e AWS002,AWS017
 
 .PHONY: test
@@ -46,8 +58,8 @@ test: ## Runs ShellSpec tests
 .PHONY: clean
 clean: ## Deletes temporary files
 	rm -rf report
-	rm jx-requirements.yml
-	rm ./tfsec
+	rm -f jx-requirements.yml
+	rm -f tfsec
 
 .PHONY: markdown-table
 markdown-table: ## Creates markdown tables for in- and output of this module

--- a/Makefile
+++ b/Makefile
@@ -28,7 +28,18 @@ lint: init ## Verifies Terraform syntax
 fmt: ## Reformats Terraform files accoring to standard
 	terraform fmt
 
-.PHONY: test 
+install-tfsec: ##Installs tfsec
+	curl -L "$$(curl -s https://api.github.com/repos/liamg/tfsec/releases/latest | grep -o -E "https://.+?-linux-amd64")" > tfsec;\
+	chmod +x ./tfsec
+
+check-tfsec: # check if tfsec is installed
+	./tfsec --version
+
+.PHONY: tfsec
+tfsec: install-tfsec check-tfsec #Runs tfsec
+	./tfsec . -e AWS002,AWS017
+
+.PHONY: test
 test: ## Runs ShellSpec tests
 	shellspec --format document --warning-as-failure
 
@@ -36,6 +47,7 @@ test: ## Runs ShellSpec tests
 clean: ## Deletes temporary files
 	rm -rf report
 	rm jx-requirements.yml
+	rm ./tfsec
 
 .PHONY: markdown-table
 markdown-table: ## Creates markdown tables for in- and output of this module

--- a/README.md
+++ b/README.md
@@ -117,6 +117,7 @@ The following sections provide a full list of configuration in- and output varia
 | create\_and\_configure\_subdomain | Flag to create an NS record set for the subdomain in the apex domain's Hosted Zone | `bool` | `false` | no |
 | desired\_node\_count | The number of worker nodes to use for the cluster | `number` | `3` | no |
 | enable\_external\_dns | Flag to enable or disable External DNS in the final `jx-requirements.yml` file | `bool` | `false` | no |
+| enable\_key\_rotation | Flag to enable kms key rotation | `bool` | `true` | no |
 | enable\_logs\_storage | Flag to enable or disable long term storage for logs | `bool` | `true` | no |
 | enable\_node\_group | Flag to enable node group | `bool` | `false` | no |
 | enable\_reports\_storage | Flag to enable or disable long term storage for reports | `bool` | `true` | no |

--- a/jenkins-x.yml
+++ b/jenkins-x.yml
@@ -30,5 +30,9 @@ pipelineConfig:
             command: make
             args:
             - lint
+          - name: terraform-security-check
+            command: make
+            args:
+            - tfsec
           - name: terraform-apply-and-test
             command: ./scripts/ci.sh

--- a/jenkins-x.yml
+++ b/jenkins-x.yml
@@ -33,6 +33,6 @@ pipelineConfig:
           - name: terraform-security-check
             command: make
             args:
-            - tfsec
+            - run-tfsec
           - name: terraform-apply-and-test
             command: ./scripts/ci.sh

--- a/modules/vault/main.tf
+++ b/modules/vault/main.tf
@@ -69,8 +69,9 @@ resource "aws_dynamodb_table" "vault-dynamodb-table" {
 // See https://www.terraform.io/docs/providers/aws/r/kms_key.html
 // ----------------------------------------------------------------------------
 resource "aws_kms_key" "kms_vault_unseal" {
-  description = "KMS Key for bank vault unseal"
-  policy      = <<POLICY
+  description         = "KMS Key for bank vault unseal"
+  enable_key_rotation = var.enable_key_rotation
+  policy              = <<POLICY
 {
     "Version": "2012-10-17",
     "Statement": [

--- a/modules/vault/variables.tf
+++ b/modules/vault/variables.tf
@@ -19,3 +19,9 @@ variable "force_destroy" {
   type        = bool
   default     = false
 }
+
+variable "enable_key_rotation" {
+  description = "Flag to enable kms key rotation"
+  type        = bool
+  default     = true
+}

--- a/scripts/ci.sh
+++ b/scripts/ci.sh
@@ -37,6 +37,9 @@ aws --version
 echo "Initializing modules..."
 terraform init
 
+echo "Validating modules ..."
+terraform validate
+
 echo "Creating cluster ${CLUSTER_NAME}"
 
 echo "Applying Terraform..."

--- a/variables.tf
+++ b/variables.tf
@@ -178,6 +178,9 @@ variable "enable_node_group" {
 
 variable "enable_worker_group" {
   description = "Flag to enable worker group"
+  type        = bool
+  default     = true
+}
 
 variable "enable_key_rotation" {
   description = "Flag to enable kms key rotation"

--- a/variables.tf
+++ b/variables.tf
@@ -178,6 +178,9 @@ variable "enable_node_group" {
 
 variable "enable_worker_group" {
   description = "Flag to enable worker group"
+
+variable "enable_key_rotation" {
+  description = "Flag to enable kms key rotation"
   type        = bool
   default     = true
 }

